composer.json

```diff
@@ -9,6 +9,10 @@
     "source": "http://cgit.drupalcode.org/rabbitmq"
   },
   "require": {
+    "ext-pcntl": "*",
```

Comment:
@FGM why do you require this ?
You don't use it anywhere in the code that extension as far I see ... ?

Reply:
It is the (core bundled) extension containing the pcntl_alarm/pcntl_signal functions which make the timeout mechanism actually work. They are used in `Consumer::consume()`.

Reply:
@FGM Why do you want to add that extension? if they stop to release/maintain it in php 7.3 would be an issue, and people that can't install/enable extensions this could be a problem

Reply:
This is need to make the timeout mechanism actually work regardless of the driver behavior. Also, this extension is bundled by default in PHP, what makes you think it could become deprecated in a future version beyond PHP 7.2 ?

Reply:
@FGM sorry I didn't mean that I know what is gonna happen in the future ... I mean completely the opposite.

Maybe is better if you split this PR into 2? one with the refactor and fixes and another one with the implementation of that extension and I can make a proper release with that change and people can discuss if that is ok or not...

I have several sites in maintenance that they don't have that extension enabled, and probably I'm not the only one with that issue

Reply:
Well, I don't have time to do this right now, preparing for DC Vienna and with the site going live next week. But I wouldn't object if someone did that split, of course. Keeping in mind that the current timeout system (included in another d.o. issue) just doesn't work in the general case; we've had lots of unexpected problems in production with the current version, and this fixes the issue by going one level lower. 

I'm very surprised some sites don't have that extension since it is bundled with PHP by default, so sites don't need to load it as they have it already. But maybe a better way would be to mark it as recommended instead of required, and wrap the pcntl-dependent calls in an availability check ?

Reply:
Also, I just noticed that the underlying driver add extra signals with that extension (see https://github.com/php-amqplib/php-amqplib/blob/master/README.md), so we could extend the handling further in a followup.

Reply:
yep, I think would be better talk after the DC, enjoy the last DC europe!!

Reply:
Hopfully not the last. But maybe the last under the auspices of the US D.A.